File: motor/motor_odom/src/odom_pub.cpp
/*
 * 2023-04-29
 * gets ticks per second from each wheel and publishes linear and angular velocity
 * from https://automaticaddison.com/how-to-publish-wheel-odometry-information-over-ros/
 * note: tutorial is in Melodic
 * 
 * subscribed to:
 *   right_wheel/ticks (ticks since last, Int32)
 *   left_wheel/ticks
 *   right_wheel/direction (bool)
 *   left_wheel/direction
 *   rover_pose/set (geometry_msgs/PoseStamped)
 *   rover_pose/reset (bool)
 * 
 * publishes to:
 *   odom (quaternion)
 * 
 * resources
 *   covariance matrix: https://answers.ros.org/question/64759/covariance-matrix-for-vo-and-odom/
 *   motion model: https://www.roboticsbook.org/S52_diffdrive_actions.html
 *  
 * 2024-04-10
 * fix direction issue - direction retrieved from command messages and
 * and combined with directionless ticks/s from hall effect sensors
 * 
 * 2024-05-20
 * subscribe to wheel/ticks instead, not necessarily ticks_ps
 * 
 * 2024-06-01
 * subscribe to rover_pose to reset position without relaunching
 * 
 * 2026-02-28
 * ROS2 Port
*/

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/int32.hpp>
#include <std_msgs/msg/float64.hpp>
#include <std_msgs/msg/bool.hpp>
#include <std_msgs/msg/string.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_ros/transform_broadcaster.h>
#include <cmath>
#include <sstream>

using namespace std;
using std::placeholders::_1;

class WheelOdomPub : public rclcpp::Node
{
public:
    WheelOdomPub() : Node("wheel_odom_pub")
    {
        // set data fields of odometry message
        odom_new_.header.frame_id = "odom";
        odom_new_.child_frame_id = "base_link";
        odom_new_.pose.pose.position.z = 0;
        odom_new_.pose.pose.orientation.x = 0;
        odom_new_.pose.pose.orientation.y = 0;
        odom_new_.twist.twist.linear.x = 0;
        odom_new_.twist.twist.linear.y = 0;
        odom_new_.twist.twist.linear.z = 0;
        odom_new_.twist.twist.angular.x = 0;
        odom_new_.twist.twist.angular.y = 0;
        odom_new_.twist.twist.angular.z = 0;
        odom_old_.pose.pose.position.x = INITIAL_X;
        odom_old_.pose.pose.position.y = INITIAL_Y;
        odom_old_.pose.pose.orientation.z = INITIAL_THETA;

        // publishers
        odom_data_pub_ = this->create_publisher<nav_msgs::msg::Odometry>(
            "odom", 100);    // full odom message, orientation as quaternion
        debug_pub_ = this->create_publisher<std_msgs::msg::String>(
            "debug_wheel_odom", 100);

        // subscribers
        // ticks per second from both wheels
        right_ticks_sub_ = this->create_subscription<std_msgs::msg::Int32>(
            "/right_wheel/ticks", 100,
            std::bind(&WheelOdomPub::right_ticks_cb, this, _1));
        left_ticks_sub_ = this->create_subscription<std_msgs::msg::Int32>(
            "/left_wheel/ticks", 100,
            std::bind(&WheelOdomPub::left_ticks_cb, this, _1));
        // wheel commands to get direction for both wheels
        right_dir_sub_ = this->create_subscription<std_msgs::msg::Bool>(
            "/right_wheel/direction", 100,
            std::bind(&WheelOdomPub::right_direction_cb, this, _1));
        left_dir_sub_ = this->create_subscription<std_msgs::msg::Bool>(
            "/left_wheel/direction", 100,
            std::bind(&WheelOdomPub::left_direction_cb, this, _1));
        // force set rover position
        pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
            "rover_pose/set", 1,
            std::bind(&WheelOdomPub::set_pose_cb, this, _1));
        reset_sub_ = this->create_subscription<std_msgs::msg::Bool>(
            "rover_pose/reset", 1,
            std::bind(&WheelOdomPub::reset_pose_cb, this, _1));

        // timer at 30 Hz
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(33),
            std::bind(&WheelOdomPub::timer_callback, this));
    }

private:
    // constants
    static constexpr double INITIAL_X = 0.0;
    static constexpr double INITIAL_Y = 0.0;
    static constexpr double INITIAL_THETA = 0.00000000001;
    static constexpr double PI = 3.1415926;

    // Approximately 9.8inches = 24.892cm diameter (estimate)
    static constexpr double WHEEL_RADIUS = 0.125; // (in metres)
    static constexpr double CIRCUMFERENCE = 2 * PI * WHEEL_RADIUS;
    static constexpr double WHEEL_BASE = 0.69; // (centre of left tire to centre of right tire)

    // defines slope of rpm/(ticks per second) reading, with intercept set to 0
    static constexpr double A = 0.3114;
    static constexpr double TICKS_PER_METRE = 1.0 / A / CIRCUMFERENCE * 60.0;

    // member variables
    nav_msgs::msg::Odometry odom_new_;
    nav_msgs::msg::Odometry odom_old_;

    // distance both wheels have travelled
    double distance_left_ = 0;
    double distance_right_ = 0;

    // ticks for each wheel
    float ticks_left_ = 0;
    float ticks_right_ = 0;

    // direction for each wheel
    int l_direction_ = 0;
    int r_direction_ = 0;

    // publishers
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_data_pub_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr debug_pub_;

    // subscribers
    rclcpp::Subscription<std_msgs::msg::Int32>::SharedPtr right_ticks_sub_;
    rclcpp::Subscription<std_msgs::msg::Int32>::SharedPtr left_ticks_sub_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr right_dir_sub_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr left_dir_sub_;
    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_sub_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr reset_sub_;

    // timer
    rclcpp::TimerBase::SharedPtr timer_;

    // --- callbacks ---

    void right_ticks_cb(const std_msgs::msg::Int32::SharedPtr msg)
    {
        ticks_right_ = msg->data;
        distance_right_ = ticks_right_ / TICKS_PER_METRE;
    }

    void left_ticks_cb(const std_msgs::msg::Int32::SharedPtr msg)
    {
        ticks_left_ = msg->data;
        distance_left_ = ticks_left_ / TICKS_PER_METRE;
    }

    void left_direction_cb(const std_msgs::msg::Bool::SharedPtr msg)
    {
        l_direction_ = msg->data ? 1 : -1;
    }

    void right_direction_cb(const std_msgs::msg::Bool::SharedPtr msg)
    {
        r_direction_ = msg->data ? 1 : -1;
    }

    void set_pose_cb(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
    {
        odom_old_.pose.pose.position.x = msg->pose.position.x;
        odom_old_.pose.pose.position.y = msg->pose.position.y;
        odom_old_.pose.pose.orientation.z = msg->pose.orientation.z;
    }

    void reset_pose_cb(const std_msgs::msg::Bool::SharedPtr msg)
    {
        if (msg->data) {
            odom_old_.pose.pose.position.x = 0;
            odom_old_.pose.pose.position.y = 0;
            odom_old_.pose.pose.orientation.z = 0;
        }
    }

    // --- timer callback (replaces while loop) ---

    void timer_callback()
    {
        update_odom();
        publish_quat();
        RCLCPP_DEBUG(this->get_logger(), "left_dir: %d\tright_dir: %d",
                     l_direction_, r_direction_);
    }

    // --- odometry update ---

    void update_odom()
    {
        // average distance since last cycle
        double cycle_distance = ((r_direction_ * distance_right_) + (l_direction_ * distance_left_)) / 2;
        // number of radians the robot has turned since the last cycle
        double cycle_angle = asin(((l_direction_ * distance_left_) - (r_direction_ * distance_right_)) / WHEEL_BASE);
        // average angle during the last cycle
        double avg_angle = cycle_angle / 2 + odom_old_.pose.pose.orientation.z;

        if (avg_angle > PI) {
            avg_angle -= 2 * PI;
        } else if (avg_angle < -PI) {
            avg_angle += 2 * PI;
        }

        // calculate new pose
        odom_new_.pose.pose.position.x = odom_old_.pose.pose.position.x + cos(avg_angle) * cycle_distance;
        odom_new_.pose.pose.position.y = odom_old_.pose.pose.position.y + sin(avg_angle) * cycle_distance;
        odom_new_.pose.pose.orientation.z = cycle_angle + odom_old_.pose.pose.orientation.z;

        // prevent lockup from a single bad cycle
        if (isnan(odom_new_.pose.pose.position.x) || isnan(odom_new_.pose.pose.position.y)
            || isnan(odom_new_.pose.pose.position.z)) {
            odom_new_.pose.pose.position.x = odom_old_.pose.pose.position.x;
            odom_new_.pose.pose.position.y = odom_old_.pose.pose.position.y;
            odom_new_.pose.pose.orientation.z = odom_old_.pose.pose.orientation.z;
        }

        // ensure theta stays in the correct range
        if (odom_new_.pose.pose.orientation.z > PI) {
            odom_new_.pose.pose.orientation.z -= 2 * PI;
        } else if (odom_new_.pose.pose.orientation.z < -PI) {
            odom_new_.pose.pose.orientation.z += 2 * PI;
        }

        // compute velocity
        odom_new_.header.stamp = this->now();
        double dt = (rclcpp::Time(odom_new_.header.stamp) - rclcpp::Time(odom_old_.header.stamp)).seconds();
        if (dt > 0.0) {
            odom_new_.twist.twist.linear.x = cycle_distance / dt;
            odom_new_.twist.twist.angular.z = cycle_angle / dt;
        }

        // save pose data for next cycle
        odom_old_.pose.pose.position.x = odom_new_.pose.pose.position.x;
        odom_old_.pose.pose.position.y = odom_new_.pose.pose.position.y;
        odom_old_.pose.pose.orientation.z = odom_new_.pose.pose.orientation.z;
        odom_old_.header.stamp = odom_new_.header.stamp;

        // publish is handled in publish_quat()
    }

    // --- publish quaternion odometry ---

    void publish_quat()
    {
        tf2::Quaternion q;
        q.setRPY(0, 0, odom_new_.pose.pose.orientation.z);

        nav_msgs::msg::Odometry quat_odom;
        quat_odom.header.stamp = odom_new_.header.stamp;
        quat_odom.header.frame_id = "odom";
        quat_odom.child_frame_id = "base_link";
        quat_odom.pose.pose.position.x = odom_new_.pose.pose.position.x;
        quat_odom.pose.pose.position.y = odom_new_.pose.pose.position.y;
        quat_odom.pose.pose.position.z = odom_new_.pose.pose.position.z;
        quat_odom.pose.pose.orientation.x = q.x();
        quat_odom.pose.pose.orientation.y = q.y();
        quat_odom.pose.pose.orientation.z = q.z();
        quat_odom.pose.pose.orientation.w = q.w();
        quat_odom.twist.twist.linear.x = odom_new_.twist.twist.linear.x;
        quat_odom.twist.twist.linear.y = odom_new_.twist.twist.linear.y;
        quat_odom.twist.twist.linear.z = odom_new_.twist.twist.linear.z;
        quat_odom.twist.twist.angular.x = odom_new_.twist.twist.angular.x;
        quat_odom.twist.twist.angular.y = odom_new_.twist.twist.angular.y;
        quat_odom.twist.twist.angular.z = odom_new_.twist.twist.angular.z;

        // build covariance matrix (use big number if unsure of uncertainty)
        for (int i = 0; i < 36; i++) {
            if (i == 0 || i == 7 || i == 14) {
                quat_odom.pose.covariance[i] = 0.01;    // translation accuracy +/- 0.01 m
            } else if (i == 21 || i == 28 || i == 35) {
                quat_odom.pose.covariance[i] = 0.1;     // rotation accuracy +/- 0.1 radian
            } else {
                quat_odom.pose.covariance[i] = 0;
            }
        }

        odom_data_pub_->publish(quat_odom);
    }
};

// main
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<WheelOdomPub>());
    rclcpp::shutdown();
    return 0;
}

File: motor/motor_odom/src/odom_pub.py
#!/usr/bin/env python3
"""
Wheel Odometry Publisher
========================
Computes odometry from wheel tick counts and publishes as nav_msgs/Odometry.

Subscribed to:
  /right_wheel/ticks  (std_msgs/Int32)
  /left_wheel/ticks   (std_msgs/Int32)
  /right_wheel/direction (std_msgs/Bool)
  /left_wheel/direction  (std_msgs/Bool)
  rover_pose/set   (geometry_msgs/PoseStamped)
  rover_pose/reset (std_msgs/Bool)

Publishes:
  odom (nav_msgs/Odometry)

History:
  2023-04-29  Initial version (C++)
  2024-04-10  Direction fix
  2024-05-20  Subscribe to wheel/ticks
  2024-06-01  Pose reset support
  2026-02-28  ROS2 Port (C++)
  2026-04-20  Python port
"""

import math

import rclpy
from rclpy.node import Node

from std_msgs.msg import Int32, Bool, String
from nav_msgs.msg import Odometry
from geometry_msgs.msg import PoseStamped, Quaternion


def quaternion_from_yaw(yaw: float) -> Quaternion:
    """Convert a yaw angle (radians) to a Quaternion message."""
    q = Quaternion()
    q.x = 0.0
    q.y = 0.0
    q.z = math.sin(yaw / 2.0)
    q.w = math.cos(yaw / 2.0)
    return q


class WheelOdomPub(Node):

    # --- Constants ---
    INITIAL_X = 0.0
    INITIAL_Y = 0.0
    INITIAL_THETA = 1e-11  # tiny offset to avoid division by zero

    PI = math.pi
    WHEEL_RADIUS = 0.1375          # metres (~9.8 in / 24.9 cm diameter)
    CIRCUMFERENCE = 2.0 * PI * WHEEL_RADIUS
    WHEEL_BASE = 0.69             # centre-to-centre of left/right tyres

    A = 0.3114                    # slope of rpm / (ticks per second)
    TICKS_PER_METRE = 1.0 / A / CIRCUMFERENCE * 60.0

    def __init__(self):
        super().__init__('wheel_odom_pub')

        # --- State ---
        self._ticks_left = 0.0
        self._ticks_right = 0.0
        self._distance_left = 0.0
        self._distance_right = 0.0
        self._l_direction = 0
        self._r_direction = 0

        # Current / previous odometry (store only the scalars we need)
        self._x = self.INITIAL_X
        self._y = self.INITIAL_Y
        self._theta = self.INITIAL_THETA
        self._prev_x = self.INITIAL_X
        self._prev_y = self.INITIAL_Y
        self._prev_theta = self.INITIAL_THETA
        self._prev_stamp = self.get_clock().now()

        self._vx = 0.0
        self._wz = 0.0

        # --- Publishers ---
        self._odom_pub = self.create_publisher(Odometry, 'odom', 100)
        self._debug_pub = self.create_publisher(String, 'debug_wheel_odom', 100)

        # --- Subscribers ---
        self.create_subscription(Int32, '/right_wheel/ticks', self._right_ticks_cb, 100)
        self.create_subscription(Int32, '/left_wheel/ticks', self._left_ticks_cb, 100)
        self.create_subscription(Bool, '/right_wheel/direction', self._right_dir_cb, 100)
        self.create_subscription(Bool, '/left_wheel/direction', self._left_dir_cb, 100)
        self.create_subscription(PoseStamped, 'rover_pose/set', self._set_pose_cb, 1)
        self.create_subscription(Bool, 'rover_pose/reset', self._reset_pose_cb, 1)

        # --- 30 Hz timer ---
        self.create_timer(1.0 / 30.0, self._timer_cb)

    # ------------------------------------------------------------------
    # Callbacks
    # ------------------------------------------------------------------

    def _right_ticks_cb(self, msg: Int32):
        self._ticks_right = float(msg.data)
        self._distance_right = self._ticks_right / self.TICKS_PER_METRE

    def _left_ticks_cb(self, msg: Int32):
        self._ticks_left = float(msg.data)
        self._distance_left = self._ticks_left / self.TICKS_PER_METRE

    def _right_dir_cb(self, msg: Bool):
        self._r_direction = 1 if msg.data else -1

    def _left_dir_cb(self, msg: Bool):
        self._l_direction = 1 if msg.data else -1

    def _set_pose_cb(self, msg: PoseStamped):
        self._prev_x = msg.pose.position.x
        self._prev_y = msg.pose.position.y
        self._prev_theta = msg.pose.orientation.z

    def _reset_pose_cb(self, msg: Bool):
        if msg.data:
            self._prev_x = 0.0
            self._prev_y = 0.0
            self._prev_theta = 0.0

    # ------------------------------------------------------------------
    # Timer
    # ------------------------------------------------------------------

    def _timer_cb(self):
        self._update_odom()
        self._publish_odom()
        self.get_logger().debug(
            f'left_dir: {self._l_direction}\tright_dir: {self._r_direction}')

    # ------------------------------------------------------------------
    # Odometry math
    # ------------------------------------------------------------------

    def _update_odom(self):
        # Average distance since last cycle
        cycle_distance = (
            (self._r_direction * self._distance_right)
            + (self._l_direction * self._distance_left)
        ) / 2.0

        # Radians turned since last cycle
        diff = (
            (self._r_direction * self._distance_right)
            - (self._l_direction * self._distance_left)
        )
        cycle_angle = math.asin(max(-1.0, min(1.0, diff / self.WHEEL_BASE)))

        # Average angle during the cycle
        avg_angle = cycle_angle / 2.0 + self._prev_theta
        if avg_angle > self.PI:
            avg_angle -= 2.0 * self.PI
        elif avg_angle < -self.PI:
            avg_angle += 2.0 * self.PI

        # New pose
        new_x = self._prev_x + math.cos(avg_angle) * cycle_distance
        new_y = self._prev_y + math.sin(avg_angle) * cycle_distance
        new_theta = cycle_angle + self._prev_theta

        # Guard against NaN from a bad cycle
        if math.isnan(new_x) or math.isnan(new_y) or math.isnan(new_theta):
            new_x = self._prev_x
            new_y = self._prev_y
            new_theta = self._prev_theta

        # Wrap theta to [-pi, pi]
        if new_theta > self.PI:
            new_theta -= 2.0 * self.PI
        elif new_theta < -self.PI:
            new_theta += 2.0 * self.PI

        # Velocity
        now = self.get_clock().now()
        dt = (now - self._prev_stamp).nanoseconds * 1e-9
        if dt > 0.0:
            self._vx = cycle_distance / dt
            self._wz = cycle_angle / dt

        # Store for next cycle
        self._x = new_x
        self._y = new_y
        self._theta = new_theta
        self._prev_x = new_x
        self._prev_y = new_y
        self._prev_theta = new_theta
        self._prev_stamp = now

    def _publish_odom(self):
        q = quaternion_from_yaw(self._theta)

        odom = Odometry()
        odom.header.stamp = self._prev_stamp.to_msg()
        odom.header.frame_id = 'odom'
        odom.child_frame_id = 'base_link'

        odom.pose.pose.position.x = self._x
        odom.pose.pose.position.y = self._y
        odom.pose.pose.position.z = 0.0
        odom.pose.pose.orientation = q

        odom.twist.twist.linear.x = self._vx
        odom.twist.twist.angular.z = self._wz

        # Covariance: diagonal entries only
        cov = [0.0] * 36
        cov[0] = 0.01    # x
        cov[7] = 0.01    # y
        cov[14] = 0.01   # z
        cov[21] = 0.1    # roll
        cov[28] = 0.1    # pitch
        cov[35] = 0.1    # yaw
        odom.pose.covariance = cov

        self._odom_pub.publish(odom)


def main(args=None):
    rclpy.init(args=args)
    node = WheelOdomPub()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
